AppAuth.podspec

```diff
@@ -33,16 +33,25 @@ It follows the OAuth 2.0 for Native Apps best current practice
   s.platforms    = { :ios => "7.0", :osx => "10.9", :watchos => "2.0", :tvos => "9.0" }
 
   s.source       = { :git => "https://github.com/openid/AppAuth-iOS.git", :tag => s.version }
-
-  s.source_files = "Source/*.{h,m}"
   s.requires_arc = true
 
-  # iOS
-  s.ios.source_files      = "Source/iOS/**/*.{h,m}"
-  s.ios.deployment_target = "7.0"
-  s.ios.frameworks        = "SafariServices", "AuthenticationServices"
-
-  # macOS
-  s.osx.source_files = "Source/macOS/**/*.{h,m}"
-  s.osx.deployment_target = '10.9'
+  # Subspec for the core AppAuth library classes only, suitable for extensions.
+  s.subspec 'Core' do |core|
+     core.source_files = "Source/*.{h,m}"
+  end
+
+  # Subspec for the full AppAuth library, including platform-dependant external user agents.
+  s.subspec 'ExternalUserAgent' do |externalUserAgent|
+    externalUserAgent.dependency 'AppAuth/Core'
+    externalUserAgent.xcconfig = { 'GCC_PREPROCESSOR_DEFINITIONS' => %w($(inherited) APPAUTH_EXTERNALUSERAGENT=1) }
```

Comment:
This line is causing me problems at `pod install`. I get this error:
`NoMethodError - undefined method `strip' for ["$(inherited)", "APPAUTH_EXTERNALUSERAGENT=1"]:Array`

Then I ran `pod spec lint` on the `AppAuth.podpsec` file and I got this:
```
 -> AppAuth (1.0.0.beta1)
    - ERROR | [iOS] unknown: Encountered an unknown error (undefined method `strip' for ["$(inherited)", "APPAUTH_EXTERNALUSERAGENT=1"]:Array) during validation.

Analyzed 1 podspec.

[!] The spec did not pass validation, due to 1 error.
```

Reply:
Hmm, `pod install` works for me. What version of CocoaPods are you using?

Mine is 1.6.0-beta.1

Reply:
Oh ok I'm not using the beta, still on the last release 1.5.3.

Reply:
I'll try your version.

Reply:
By replacing the line in error by the following it seems to work:
```
externalUserAgent.xcconfig = { 'GCC_PREPROCESSOR_DEFINITIONS' => '$(inherited) APPAUTH_EXTERNALUSERAGENT=1' }
```

The only thing I noted is that it adds duplicate `$(inherited)` in the `AppAuth.xcconfig` file.
```
GCC_PREPROCESSOR_DEFINITIONS = $(inherited) COCOAPODS=1 $(inherited) APPAUTH_EXTERNALUSERAGENT=1
```
Then based on that, removing `$(inherited)` from the podspec should be fine.

Also should we use the subspec property `pod_target_xcconfig` instead of `xcconfig`. From what I understand `xcconfig` is like setting both `pod_target_xcconfig `and `user_target_xcconfig`. Setting `user_target_xcconfig` is not recommended as per [documentation](https://guides.cocoapods.org/syntax/podspec.html#user_target_xcconfig) and which I guess is not needed here.

Reply:
I can confirm that replacing that line in the `.podspec` fixes the issue with `pod install`. But then after I installed the pod and build & run `Example-iOS_ObjC` and press the "Authorize" button I see the exact same issue as I had in my initial PR:

```
2018-10-18 12:53:26.448073+0200 Example-iOS_ObjC[20935:3600206] *** Assertion failure in -[AppAuthExampleViewController verifyConfig], /Users/tschroep/Documents/Personal/git/github.com/williamdeniss/AppAuth-iOS/Examples/Example-iOS_ObjC/Source/AppAuthExampleViewController.m:76
2018-10-18 12:53:26.461230+0200 Example-iOS_ObjC[20935:3600206] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Update kIssuer with your own issuer. Instructions: https://github.com/openid/AppAuth-iOS/blob/master/Examples/Example-iOS_ObjC/README.md'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010503429b __exceptionPreprocess + 331
	1   libobjc.A.dylib                     0x00000001045d0735 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000105034022 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x0000000103fd4b47 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 194
	4   Example-iOS_ObjC                    0x00000001039b348d -[AppAuthExampleViewController verifyConfig] + 301
	5   Example-iOS_ObjC                    0x00000001039b56fb -[AppAuthExampleViewController authWithAutoCodeExchange:] + 59
	6   UIKitCore                           0x0000000105c0c7c3 -[UIApplication sendAction:to:from:forEvent:] + 83
	7   UIKitCore                           0x0000000105d44e85 -[UIControl sendAction:to:forEvent:] + 67
	8   UIKitCore                           0x0000000105d451a2 -[UIControl _sendActionsForEvents:withEvent:] + 450
	9   UIKitCore                           0x0000000105d440e6 -[UIControl touchesEnded:withEvent:] + 583
	10  UIKitCore                           0x000000010641f334 -[UIWindow _sendTouchesForEvent:] + 2729
	11  UIKitCore                           0x0000000106420a30 -[UIWindow sendEvent:] + 4080
	12  UIKitCore                           0x0000000105c26e10 -[UIApplication sendEvent:] + 352
	13  UIKit                               0x00000001273b94eb -[UIApplicationAccessibility sendEvent:] + 85
	14  UIKitCore                           0x0000000105b5f0d0 __dispatchPreprocessedEventFromEventQueue + 3024
	15  UIKitCore                           0x0000000105b61cf2 __handleEventQueueInternal + 5948
	16  CoreFoundation                      0x0000000104f97b31 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	17  CoreFoundation                      0x0000000104f973a3 __CFRunLoopDoSources0 + 243
	18  CoreFoundation                      0x0000000104f91a4f __CFRunLoopRun + 1263
	19  CoreFoundation                      0x0000000104f91221 CFRunLoopRunSpecific + 625
	20  GraphicsServices                    0x000000010d71b1dd GSEventRunModal + 62
	21  UIKitCore                           0x0000000105c0b115 UIApplicationMain + 140
	22  Example-iOS_ObjC                    0x00000001039b3120 main + 112
	23  libdyld.dylib                       0x0000000109e35551 start + 1
```

This was the reason why I stalled work on my original PR and created the second PR which adds the `AppAuthCore.podspec`.
@WilliamDenniss and @julienbodet can you confirm this issue or is it just an issue in my local setup?

Reply:
I can confirm that @julienbodet's change fixes the issue with `pod install`. But now if I build & run my project with this PR included I get the following error (at runtime!): 

```
2018-10-18 12:51:58.487329+0200 XXXXX[20739:3597215] +[OIDAuthState authStateByPresentingAuthorizationRequest:presentingViewController:callback:]: unrecognized selector sent to class 0x1051a79f0
```

This is the same error I had with my initial commit which was the reason why I suspended work on the initial PR and created the second PR which introduced the new `AppAuthCore.podspec`.
Currently I don't have the example apps properly configured so I can't reproduce this issue there yet.
@WilliamDenniss and @julienbodet can you confirm this issue? or is it a problem in my local setup?

Reply:
did you also test a full login (i.e. was the `[OIDAuthState authStateByPresentingAuthorizationRequest:presentingViewController:callback:]` method actually invoked in your test and the login screen displayed correctly)?

Reply:
@schroepf Yes the safari instance presented the login screen and I then I got the callback once the login process was completed.

Reply:
The pod is definitely working for me, but I can't get it to pass `pod spec lint`, it errors testing the module import. 

```
    CompileC /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/Objects-normal/i386/main.o /var/folders/fb/9vm1n3mj2mq40j5jrjd4pmgh007sbp/T/CocoaPods-Lint-20181017-14224-1pmk0me-AppAuth/App/main.m normal i386 objective-c com.apple.compilers.llvm.clang.1_0.compiler (in target: App)
        cd /var/folders/fb/9vm1n3mj2mq40j5jrjd4pmgh007sbp/T/CocoaPods-Lint-20181017-14224-1pmk0me-AppAuth
        export LANG=en_US.US-ASCII
        /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x objective-c -arch i386 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=gnu11 -fobjc-arc -fobjc-weak -fmodules -gmodules -fmodules-cache-path=/Users/wdenniss/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -fmodules-prune-interval=86400 -fmodules-prune-after=345600 -fbuild-session-file=/Users/wdenniss/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror=non-modular-include-in-framework-module -Wno-trigraphs -fpascal-strings -Os -fno-common -Wno-missing-field-initializers -Wno-missing-prototypes -Werror=return-type -Wdocumentation -Wunreachable-code -Wno-implicit-atomic-properties -Werror=deprecated-objc-isa-usage -Wno-objc-interface-ivars -Werror=objc-root-class -Wno-arc-repeated-use-of-weak -Wimplicit-retain-self -Wduplicate-method-match -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-float-conversion -Wnon-literal-null-conversion -Wobjc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -Wno-selector -Wno-strict-selector-match -Wundeclared-selector -Wdeprecated-implementations -DCOCOAPODS=1 -DNS_BLOCK_ASSERTIONS=1 -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator12.0.sdk -fasm-blocks -fstrict-aliasing -Wprotocol -Wdeprecated-declarations -mios-simulator-version-min=8.0 -g -fvisibility=hidden -Wno-sign-conversion -Winfinite-recursion -Wcomma -Wblock-capture-autoreleasing -Wstrict-prototypes -Wno-semicolon-before-method-body -Wunguarded-availability -fobjc-abi-version=2 -fobjc-legacy-dispatch -iquote /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/App-generated-files.hmap -I/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/App-own-target-headers.hmap -I/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/App-all-non-framework-target-headers.hmap -ivfsoverlay /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/all-product-headers.yaml -iquote /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/App-project-headers.hmap -I/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator/include -I/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/DerivedSources/i386 -I/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/DerivedSources -F/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator -F/Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator/AppAuth -iquote /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator/AppAuth/AppAuth.framework/Headers -MMD -MT dependencies -MF /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/Objects-normal/i386/main.d --serialize-diagnostics /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/Objects-normal/i386/main.dia -c /var/folders/fb/9vm1n3mj2mq40j5jrjd4pmgh007sbp/T/CocoaPods-Lint-20181017-14224-1pmk0me-AppAuth/App/main.m -o /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Intermediates.noindex/App.build/Release-iphonesimulator/App.build/Objects-normal/i386/main.o
    While building module 'AppAuth' imported from /var/folders/fb/9vm1n3mj2mq40j5jrjd4pmgh007sbp/T/CocoaPods-Lint-20181017-14224-1pmk0me-AppAuth/App/main.m:
    3:
    In file included from <module-includes>:1:
    In file included from /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator/AppAuth/AppAuth.framework/Headers/AppAuth-umbrella.h:13:
    /Users/wdenniss/Library/Developer/Xcode/DerivedData/App-gvmbzdzisnvohpecckknfilvenfl/Build/Products/Release-iphonesimulator/AppAuth/AppAuth.framework/Headers/AppAuth.h:
    47:9
    : fatal error: 'OIDAuthState+IOS.h' file not found
    #import "OIDAuthState+IOS.h"
            ^~~~~~~~~~~~~~~~~~~~
    1 error generated
    .
    /var/folders/fb/9vm1n3mj2mq40j5jrjd4pmgh007sbp/T/CocoaPods-Lint-20181017-14224-1pmk0me-AppAuth/App/main.m:3:9: fatal error: could not build module 'AppAuth'
    @import AppAuth;
     ~~~~~~~^~~~~~~
    2 errors generated.
 ```

@schroepf I'd like to avoid maintaining 2 separate pods, I think there is a way we can get this to work.

I'm going to try adding a new header, AppAuthCore.h just for this target, see if that helps, maybe it's bad design to have conditional imports in the header.

Reply:
@WilliamDenniss totally agree to keep it simple and prefer to have only one podspec with two subpods. If we can make it work that would be great!

Now I was able to reproduce the crash I had in my own project also in the `Example-iOS_ObjC`.
Basically what I did was
- add a new target (of type "Intents Extension")
- adjust the `Podfile to use `AppAuth/Core` for the Extension target
- add `use_frameworks!` to the `Podfile`(the crash only happens when `user_frameworks` is used)

this is the podfile used in my test:

```
abstract_target 'CommonPods' do
    use_frameworks!
    
    target 'Example-iOS_ObjC' do
        platform :ios, '9.0'
        pod 'AppAuth', :path => '../../'
    end
    
    target 'Extension' do
        platform :ios, '9.0'
        pod 'AppAuth/Core', :path => '../../'
    end
    
end
```

Reply:
I created #320 which uses dedicated and separate imports. Can you try it out?